metadata-service/auth-impl/src/main/java/com/datahub/authorization/DefaultEntitySpecResolver.java

```diff
@@ -26,7 +27,8 @@ public DefaultEntitySpecResolver(Authentication systemAuthentication, EntityClie
             new DomainFieldResolverProvider(entityClient, systemAuthentication),
             new OwnerFieldResolverProvider(entityClient, systemAuthentication),
             new DataPlatformInstanceFieldResolverProvider(entityClient, systemAuthentication),
-            new GroupMembershipFieldResolverProvider(entityClient, systemAuthentication));
+            new GroupMembershipFieldResolverProvider(entityClient, systemAuthentication),
+            new TagFieldResolverProvider(entityClient, systemAuthentication));
```

Comment:
This is an inefficient design. The urns and their aspect reads should be batched and not fetching each aspect one by one for each urn. I know this isn't part of the scope for this PR, but authentication causes latency on practically every api call so it has a high impact. Even if we don't batch urns, we should definitely batch the aspects There should be a batched resolver for domain, owner, data platform, group membership, and tags. All these aspects should be fetched in batch using the entity client's same getV2 method just using the collection as something other then a single item `Collections.singleton(`

This is a perfect opportunity to address the tech debt here by adding the extra aspect and improve performance.

Reply:
After reading the adjacent code, I don't think we are sequentially performing each of these resolvers for each request. Rather in `getFieldResolvers` (in this file) we are creating a dictionary of what query to run based on the [EntityFieldType](https://github.com/datahub-project/datahub/blob/master/metadata-auth/auth-api/src/main/java/com/datahub/authorization/EntityFieldType.java) that an incoming request has in their [filter](https://github.com/datahub-project/datahub/blob/943bb57cbcf22db12c092a7f9a30c762aa2bf6e5/metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java#L243).

See https://github.com/datahub-project/datahub/blob/943bb57cbcf22db12c092a7f9a30c762aa2bf6e5/metadata-auth/auth-api/src/main/java/com/datahub/authorization/ResolvedEntitySpec.java#L25

Reply:
That line is executing a request per field meaning separate requests for each aspect. It is using futures, but not batching.

Reply:
`fieldResolvers.get(entityFieldType)` is only getting 1 resolver and generating 1 future, which depends on the filter being analysed.

I think this is the line responsible for multiple possible requests: https://github.com/datahub-project/datahub/blob/943bb57cbcf22db12c092a7f9a30c762aa2bf6e5/metadata-service/auth-impl/src/main/java/com/datahub/authorization/PolicyEngine.java#L230

Reply:
Right, if a single policy operates on 3 fields it would benefit from fetching all three at the same time. If there are multiple policies which depend on 3 different fields, it would also benefit from fetching all three at the same time. The general idea is to fetch the required aspects in 1 call for the policies being evaluated rather then 3 subsequent calls. I am assuming the same resource is being passed to each policy and that we would not be fetching the same aspect for each policy already today.

Reply:
At this point in the code, only a single
Policy’s filter gets evaluated.

Reply:
Going down the route of evaluating all policies to check whether they apply for a given resource request is a large refactor of the policy engine. I would suggest a design document before doing anything else.

@RyanHolstien @david-leifker do we have metrics on how much the PolicyEngine is a performance bottleneck?

Reply:
What happens in this case if no policies are authored with any tag predicates on them? Does the code performance stay the same as today or degrade?

Reply:
Code performs the same as today. The field resolver for tags is not invoked.